### PR TITLE
[dask] Make output of feature contribution predictions for sparse matrices match those from sklearn estimators (fixes #3881)

### LINKED ISSUE #3881: [dask] Result shape from DaskLGBMClassifier.predict(pred_contrib=True) for CSC matrices is inconsistent with LGBMClassifier
See the discussion in https://github.com/microsoft/LightGBM/pull/3866#discussion_r566106219 for full details.

`lightgbm.dask.DaskLGBMClassifier` tries to stay as close as possible to the API of `lightgbm.sklearn.LGBMClassifier`. This feature describes one known inconsistency.

In `lightgbm.sklearn.LGBMClassifier`, for multiclass classification tasks, if you call `.predict(X, pred_contrib=True)` and X is a [`scipy.sparse.csc_matrix`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.csc_matrix.html), the result will be a list of CSC matrices, 1 per class.

In `lightgbm.dask.DaskLGBMClassifier`, for multiclass classification taks, if you call `.predict(X, pred_contrib=True)` and X is a Dask Array whose partitions are each a [`scipy.sparse.csc_matrix`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.csc_matrix.html), the result will be a Dask Array that, once `.compute()`'d, returns a [`scipy.sparse.coo_matrix`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.coo_matrix.html).

To complete this feature, try to make Dask's behavior match the behavior from `lightgbm.sklearn.LGBMClassifier`, or document why that can't / shouldn't be done.

### PR DESCRIPTION
**Pull request by jameslamb**
Fixes #3881.

Resolves this comment from https://github.com/microsoft/LightGBM/pull/4351#issue-664073860.

> however this is failing for `test_classifier_pred_contrib[multiclass-classification-scipy_csr_matrix]`

## Changes in this PR

* changes `lgb.dask.DaskLGBMClassifier.predict()` to return ~a [Dask Bag](https://docs.dask.org/en/latest/bag.html)~ a list of Dask Arrays when input data is composed of sparse matrices and `pred_contrib=True`
* changes data creation function in the unit tests for `lightgbm.dask` so that it now correctly sets local data to a scipy CSR matrix when asked to do so (previously, it would return a dense numpy array)

## How this improves `lightgbm.dask`

* makes `lgb.dask.DaskLGBMClassifier.predict()`'s behavior more consistent with the equivalent behavior from `lgb.sklearn.LGBMClassifier`
* gives #4351 a path forward, to remove a critical incompatibility (`lighgbm.dask` not working with `dask>2021.4.0`)
     - which is important because it allows removal of the pin on `dask` versions in LightGBM's tests (#4307, #4285)
* fixes a related bug in the unit tests for `lightgbm.dask`, where the local array data was always an `np.ndarray` even in cases where it should have been a scipy CSR matrix

## Background

#3000 added support for getting feature contributions (using `.predict(pred_contrib=True)`) for input data stored in sparse matrices. The design decision was made in that PR that for one specific case (multiclass classificaiont + sparse input data + `pred_contrib=True`), LightGBM should return a list of sparse matrices (one per class) instead of a single concatenated-together new matrix. Maybe this is because of https://github.com/microsoft/LightGBM/pull/3000#discussion_r432694768...I'm not sure.

> I think concatenating sparse matrices together into one from multiple is a bit more expensive than for dense data - eg you have to re-calculate the indptr

## Notes for Reviewers

When given a Dask Array, `lightgbm.dask._predict()` can use Dask's built-in `.map_blocks()` to say "call `predict(**kwargs)` on each chunk of this array, then return me a new Dask Array of all those results". However, operations where you want to call a function **that does not return an array** on each chunk of a Dask Array and then combine those results are not well supported. See the discussion in https://github.com/dask/dask/issues/7589 for an explanation of why this is a very difficult problem.

~The best I could come up with was this approach, which produces a Dask Bag where, if you call `.compute()`, you'll get a result that is identical to the equivalent non-Dask code with `lgb.sklearn.LGBMClassifier.predict()`.~

I am ok with this change going into the 3.3.0 release (#4310) even though it is a breaking change, since other things are reliant on it, since it is such a narrow part of the API (only applies if using Dask and using sparse Dask Arrays and using multiclass classification and looking for feature contributions), and since it makes the Dask interface more consistent with the scikit-learn interface. The Dask community has, generally speaking, been comfortable with breaking changes if they're made in pursuit of better consistency with scikit-learn / numpy/ scikit-learn. But will defer to your opinion on that @StrikerRUS .

### CODE REVIEW

**Comment by jameslamb** [comment]
@jmoralez whenever you have time, could you please take a look at this and let me know what you think?

Here's some simpler code you could use to test that might be a little easier to experiment with interactively than the unit tests.

```python
import dask.array as da
import numpy as np
import lightgbm as lgb
from dask import delayed
from dask.distributed import Client, LocalCluster
from lightgbm.dask import DaskLGBMClassifier
from lightgbm.sklearn import LGBMClassifier
from scipy.sparse import csc_matrix, csr_matrix
from sklearn.datasets import make_blobs

n_workers = 3
cluster = LocalCluster(n_workers=n_workers)
client = Client(cluster)
client.wait_for_workers(n_workers)

print(f"View the dashboard: {cluster.dashboard_link}")

chunk_size = 50
X, y = make_blobs(n_samples=100, centers=3, random_state=42)
rnd = np.random.RandomState(42)
dX = da.from_array(X, chunks=(chunk_size, X.shape[1])).map_blocks(csc_matrix)
dy = da.from_array(y, chunks=chunk_size)

dask_clf = DaskLGBMClassifier(n_estimators=5, num_leaves=2, tree_learner="data")
dask_clf.fit(dX, dy)

preds = dask_clf.predict(dX, pred_contrib=True)
preds_computed = preds.compute()
```

**Comment by jmoralez** [comment]
Nice idea using bags! Although I'm not 100% convinced on the type of `preds` (`dask.bag.core.Item`) and I think maybe building the bag from delayeds instead of sequence might be a better idea, since I think we're nesting delayeds here (that `.compute()` in the function looks suspicious).

I believe we can maybe get `preds` to be `List[dask.array.Array]` where each array has the same chunks as `dX` and such that `client.compute(preds)` returns `List[Future]` where each future's result type is `scipy.sparse.csr_matrix`.

I think it's really up to deciding which consistency to have, with the result type of `predict` or with the result type of `preds.compute()`

**Comment by jameslamb** [comment]
> I believe we can maybe get preds to be List[dask.array.Array] where each array has the same chunks as dX

hmmm, I couldn't think of a way to do this that didn't involve having `num_classes_` duplicate calls of `predict()`. Since each call of `predict()` on one chunk of `dX` returns a list with `num_classes_` elements.

If you have an idea about the chain of calls that could produce such an output (even if it's just pseudocode), that would help me.

I'll experiment a bit more.

**Comment by jmoralez** [comment]
I was experimenting with this in the if statement for the pred_contrib, definitely not pretty (I imported dask, dask.array and dask.bag without using compat) and `client.compute(preds)` returns futures of type COO instead of CSR, but maybe it can help you.
```python
delayed_chunks = data.to_delayed()
bag = db.from_delayed(delayed_chunks[:, 0])

predict_function = partial(
 _predict_part,
 model=model,
 raw_score=False,
 pred_proba=False,
 pred_leaf=False,
 pred_contrib=True,
 **kwargs
)

@dask.delayed
def extract(l, i):
    return l[i]

preds = bag.map_partitions(predict_function)
chunks = data.chunks[0]
out = [[] for _ in range(model.n_classes_)]
for j, partition in enumerate(preds.to_delayed()):
    for i in range(model.n_classes_):
        part = da.from_delayed(
            extract(partition, i),
            shape=(chunks[j], model.n_classes_),
            meta=data._meta
        )
        out[i].append(part)
for i in range(model.n_classes_):
     out[i] = da.concatenate(out[i])
return out
```

**Comment by jameslamb** [comment]
oooo nice, thank you! That's very helpful. I'm going to keep experimenting.

**Comment by jameslamb** [comment]
I'll move this back to a draft for now

File: python-package/lightgbm/compat.py
Comment (jameslamb) [inline python-package/lightgbm/compat.py:153]: note for reviewers: I used "Mock" instead of being consistent with the use of "Dummy" elsewhere in this file because `pycodestyle` is not happy with "Dummy" for function docstrings.

Failed lint job: https://github.com/microsoft/LightGBM/pull/4378/checks?check_run_id=2836021678

> ./python-package/lightgbm/compat.py:153 in public function `dask_array_concatenate`:
        D401: First line should be in imperative mood; try rephrasing (found 'Dummy')
./python-package/lightgbm/compat.py:157 in public function `dask_array_from_delayed`:
        D401: First line should be in imperative mood; try rephrasing (found 'Dummy')
./python-package/lightgbm/compat.py:161 in public function `dask_bag_from_delayed`:
        D401: First line should be in imperative mood; try rephrasing (found 'Dummy')

**Comment by jameslamb** [comment]
Thanks so much for your help @jmoralez ! I was really stuck until I saw your idea about using delayeds and a delaying a function to extract the relevant array.

Whenever you have time, please take a look at the new state of this branch and let me know what you think. I modified that code you provided in a way that it will now return a `List[dask.array.Array]` where each array, if computed, will be a `scipy.sparse.csr.csr_matrix`.

**Comment by jmoralez** [comment]
I think this should work for both csr and csc, right? I think that could be achieved by setting `pred_meta = data._meta`, since `meta` only takes the type of the array (numpy, csr, csc) and the dtype (float32, float64), the shape is taken from the `chunks` argument. 

Testing this although the meta for the arrays is csr once computed its actually coo, I believe this is because that's the default output format when concatenating sparse matrices. Keeping the actual format can be achieved with `scipy.sparse.vstack(blocks, format='csr')` for example. I added the following before the concatenation:

```python
if isinstance(pred_meta, ss.csr_matrix):
    concat_fn = partial(ss.vstack, format='csr')
elif isinstance(pred_meta, ss.csc_matrix):
    concat_fn = partial(ss.vstack, format='csc')
else:  # not sure here, only csr and csc allowed?
    concat_fn = ss.vstack

for i in range(num_classes):
    out[i] = delayed(concat_fn)(out[i])

return out
```

`preds[0].compute()` has the correct sparse format, however `preds` is now `List[Delayed]`.

Let me know what you think.

**Comment by jameslamb** [comment]
Thanks again for your help @jmoralez !

Yes, I think it's desirable for the output to be a list of CSR matrices if the input is a CSR matrix, and CSC matrices if the input is a CSC matrix (to match the behavior of the equivalent `lightgbm.sklearn` estimators.

I also think it's preferable to have the result be a `List[dask_Array]`, if possible, so that Dask code for postprocessing those results can use array operations that could be used on the equivalent `scipy` matrices.

And I think I was able to achieve that in the recent commits added to this branch, with a slightly modified version of the approach you suggested in https://github.com/microsoft/LightGBM/pull/4378#issuecomment-862840139 :grinning: 

I've added new unit tests to confirm that CSR and CSC matrices are treated differently and correctly for this case (multiclass classification + `pred_contrib = True`), but you can also see this with the following easier-to-experiment-with code.

```python
import dask.array as da
import numpy as np

from dask.distributed import Client, LocalCluster
from lightgbm.dask import DaskLGBMClassifier
from scipy.sparse import csc_matrix, csr_matrix
from sklearn.datasets import make_blobs

n_workers = 3
cluster = LocalCluster(n_workers=n_workers)
client = Client(cluster)
client.wait_for_workers(n_workers)

print(f"View the dashboard: {cluster.dashboard_link}")

chunk_size = 50
X, y = make_blobs(n_samples=100, centers=3, random_state=42)
rnd = np.random.RandomState(42)
data = da.from_array(X, chunks=(chunk_size, X.shape[1])).map_blocks(csc_matrix)
dy = da.from_array(y, chunks=chunk_size)

dask_clf = DaskLGBMClassifier(n_estimators=5, num_leaves=2, tree_learner="data")
dask_clf.fit(data, dy)

# CSR
X_pred = da.from_array(X, chunks=(chunk_size, X.shape[1])).map_blocks(csr_matrix)
preds = dask_clf.predict(X_pred, pred_contrib=True)

print("--- X_pred ---")
print(X_pred)

print("--- preds ---")
print(preds)

print("--- computed_preds ---")
computed_preds = [p.compute() for p in preds]
print(computed_preds)

# CSC
X_pred = da.from_array(X, chunks=(chunk_size, X.shape[1])).map_blocks(csc_matrix)
preds = dask_clf.predict(X_pred, pred_contrib=True)

print("--- X_pred ---")
print(X_pred)

print("--- preds ---")
print(preds)

print("--- computed_preds ---")
computed_preds = [p.compute() for p in preds]
print(computed_preds)
```

CSR output:

> --- X_pred ---
dask.array<csr_matrix, shape=(100, 2), dtype=float64, chunksize=(50, 2), chunktype=scipy.csr_matrix>
--- preds ---
[dask.array<from-value, shape=(100, 3), dtype=float64, chunksize=(100, 3), chunktype=scipy.csr_matrix>, dask.array<from-value, shape=(100, 3), dtype=float64, chunksize=(100, 3), chunktype=scipy.csr_matrix>, dask.array<from-value, shape=(100, 3), dtype=float64, chunksize=(100, 3), chunktype=scipy.csr_matrix>]
--- computed_preds ---
[<100x3 sparse matrix of type '<class 'numpy.float64'>'
	with 200 stored elements in Compressed Sparse Row format>, <100x3 sparse matrix of type '<class 'numpy.float64'>'
	with 200 stored elements in Compressed Sparse Row format>, <100x3 sparse matrix of type '<class 'numpy.float64'>'
	with 200 stored elements in Compressed Sparse Row format>]

CSC output

> --- X_pred ---
dask.array<csc_matrix, shape=(100, 2), dtype=float64, chunksize=(50, 2), chunktype=scipy.csc_matrix>
--- preds ---
[dask.array<from-value, shape=(100, 3), dtype=float64, chunksize=(100, 3), chunktype=scipy.csc_matrix>, dask.array<from-value, shape=(100, 3), dtype=float64, chunksize=(100, 3), chunktype=scipy.csc_matrix>, dask.array<from-value, shape=(100, 3), dtype=float64, chunksize=(100, 3), chunktype=scipy.csc_matrix>]
--- computed_preds ---
[<100x3 sparse matrix of type '<class 'numpy.float64'>'
	with 200 stored elements in Compressed Sparse Column format>, <100x3 sparse matrix of type '<class 'numpy.float64'>'
	with 200 stored elements in Compressed Sparse Column format>, <100x3 sparse matrix of type '<class 'numpy.float64'>'
	with 200 stored elements in Compressed Sparse Column format>]

<hr>

Could you please take another look and let me know what you think?

**Comment by jmoralez** [comment]
Hi @jameslamb. I think we can keep the chunks as well by using `dask.array.core.concatenate.lookup_register` as described in the second point [here](https://docs.dask.org/en/latest/array-sparse.html#requirements). So adding the following somewhere:

```python
# the concat function needs an axis argument, hence this function
def sparse_vstack(blocks, format, axis=0):
    return ss.vstack(blocks, format)

da.core.concatenate.lookup_register(ss.csc_matrix, partial(sparse_vstack, format='csc'))
da.core.concatenate.lookup_register(ss.csr_matrix, partial(sparse_vstack, format='csr'))
```
And then we can just do
```python
return [da.concatenate(arr) for arr in out]
```
right after splitting the arrays into the `out` list.

I think your last version is working very well, I don't know if its really worth it keeping the chunks as well, just wanted to try it.

**Comment by jameslamb** [comment]
Thanks, I wasn't aware of the `lookup_register()` stuff. Seems like its use in `dask.array.concatenate()` was added in July 2020: https://github.com/dask/dask/pull/6339.

To be honest, I don't think it's critical that the chunks be preserved. As long as the ordering is preserved and the results are otherwise correct (which I believe is true on the current state of this PR), I think that's enough for now and we could take that additional work as a feature request.

Because this PR is necessary to unblock #4351 and #4351 is necessary to make `lightgbm.dask` compatible with newer versions of `dask`, I'm going to take this out of draft and ask @StrikerRUS to review.

Thanks again for all your help with this one!

File: tests/python_package_test/test_dask.py
Comment (StrikerRUS) [inline tests/python_package_test/test_dask.py:414]:
```suggestion
                np.testing.assert_array_equal(pred_classes, local_pred_classes)
```

**Review by StrikerRUS — COMMENTED** [review]
Nice work! Generally LGTM, just some minor comments.

File: python-package/lightgbm/dask.py
Comment (jameslamb) [reply python-package/lightgbm/dask.py:893]: Oh I did `getattr()` because I thought that `.n_classes_` and `._n_classes` would not be defined if using `LGBMRanker` or `LGBMRegressor`.

But I can see now that I was wrong about that and that `self._n_classes` is initialized to `None` in the constructor of `LGBMModel`.

Thanks for pointing it out!

File: python-package/lightgbm/dask.py
Comment (jameslamb) [reply python-package/lightgbm/dask.py:898]: excellent point! I didn't think about that.

File: python-package/lightgbm/dask.py
Comment (jameslamb) [reply python-package/lightgbm/dask.py:925]: why do you have a preference for `[]` instead of `list()` here? Are they not identical?

Comment (StrikerRUS) [reply python-package/lightgbm/dask.py:925]: https://stackoverflow.com/a/2972369

Comment (jameslamb) [reply python-package/lightgbm/dask.py:925]: ah I didn't know that, intersting!

File: tests/python_package_test/test_dask.py
Comment (jameslamb) [reply tests/python_package_test/test_dask.py:412]: I'm not sure, was just copying the existing pattern from https://github.com/microsoft/LightGBM/blob/bc9d34e4e651b5744b29a556a2e9d2301707e35b/tests/python_package_test/test_dask.py#L588. Maybe it's not needed, I'll test

Comment (jameslamb) [reply tests/python_package_test/test_dask.py]: seems like it isn't necessary! I just tested locally.

removed these (and other existing uses in `test_classifier_pred_contrib` and `test_regressor_pred_contrib`) in https://github.com/microsoft/LightGBM/pull/4378/commits/f79ab3c51cea475f97b257855aa8d9bf057685a9

File: python-package/lightgbm/dask.py
Comment (jameslamb) [reply python-package/lightgbm/dask.py:905]: ah yes I think you're right about that, this was a mistake left over from when I was developing.

**Comment by jameslamb** [comment]
Thanks for the thorough review @StrikerRUS ! I learned a few interesting things about Python and the core of `lightgbm` from it...really, really helpful.

I think I've addressed all of your comments and that this is ready for another review.

I've also created #4438, a feature request for the follow-up work suggested by https://github.com/microsoft/LightGBM/pull/4378#issuecomment-867270098.

**Review by StrikerRUS — APPROVED** [review]
Thank you very much! LGTM!

**Comment by jameslamb** [comment]
All CI jobs except one passed. `QEMU_multiarch bdist` failed with this error:

> E       ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()

https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build/results?buildId=10497&view=logs&j=c2f9361f-3c13-57db-3206-cee89820d5e3&t=bf5b33f9-9072-549e-d8b1-cab79a1dd2a9

Looks like that is coming from https://github.com/microsoft/LightGBM/blob/26cc160abc92d18a7373df538b4bbbaa1628e908/tests/python_package_test/test_dask.py#L406

I think only that QEMU job was failing because different versions of `numpy` will treat that case slightly differently. Anyway, it's a mistake in the test (a misplaced parenthesis).

I fixed that issue in https://github.com/microsoft/LightGBM/pull/4378/commits/95ae45f879942191f94e1fceaab325c6bf737e21.

**Comment by jameslamb** [comment]
ah! This test just failed again!

https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build/results?buildId=10497&view=logs&j=c2f9361f-3c13-57db-3206-cee89820d5e3&t=bf5b33f9-9072-549e-d8b1-cab79a1dd2a9

It seems the problem is just about about the use of `len()` on a numpy array, but how `np.unique()` works in the newest version of `numpy`.

I'm able to reproduce the problem on my Mac by upgrading to the newest `numpy` (v1.2.1) and `scipy` (1.7.0)

```shell
pip install --upgrade numpy scipy
cd python-package
python setup.py install
cd ../tests/python_package_test
# manually comment out "if not platform.startswith('linux')"
pytest test_dask.py::test_classifier_pred_contrib
```

I'll investigate this later tonight.

**Comment by jameslamb** [comment]
@StrikerRUS I was able to reproduce this tonight! I believe there was a breaking change in `numpy` 1.21.0 (released June 22, 2021), which causes the error I saw in https://github.com/microsoft/LightGBM/pull/4378#issuecomment-873655543.

I've opened https://github.com/numpy/numpy/issues/19405 to report it.

That issue is specific to using `np.unique()` on a `np.matrix` object. I learned that calling `.todense()` on a sparse matrix returns a `np.matrix`, not a `np.array`.

We didn't hit this issue before in LightGBM's tests because of the uses of `np.array()` to wrap results of `.todense()` (removed in https://github.com/microsoft/LightGBM/pull/4378#discussion_r663407796).

Given that, I think we should just restore those uses of `np.array()`. They don't introduce that much overhead and they allow these tests to not be blocked by this possible regression in `numpy`.

I restored them in https://github.com/microsoft/LightGBM/pull/4378/commits/831927bb9fb501a71d5a93558c6a7c2dda20770f.

**Comment by rkern** [comment]
As mentioned on the numpy issue, this is not so much of a regression. `np.unique()` never worked correctly on `np.matrix` objects. The fact that it raises an exception in one version and doesn't in an older version is not a regression (arguably, an enhancement, though both are basically accidents of never thinking about `np.matrix` in the implementation). Coercing to true `ndarray`s was always the right idea.

**Comment by rkern** [comment]
Consider using `np.asarray()`, which will not copy the underlying memory.

**Comment by jameslamb** [comment]
> Coercing to true ndarrays was always the right idea.
> Consider using `np.asarray()`, which will not copy the underlying memory.

Thanks very much for taking the time to come visit this PR and share some advice with us, @rkern! We'll definitely do that here.

**Comment by rkern** [comment]
There's also [`toarray()`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.spmatrix.toarray.html) on the `scipy.sparse` matrix objects, which bypasses `np.matrix` entirely. That would be the most efficient and idiomatic way to write that.

**Comment by jameslamb** [comment]
> There's also [`toarray()`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.spmatrix.toarray.html) on the `scipy.sparse` matrix objects, which bypasses `np.matrix` entirely. That would be the most efficient and idiomatic way to write that.

ah, even better! I'll make that change here. Thanks again for your help.

**Comment by jameslamb** [comment]
ok switching from `.todense()` to `.toarray()` worked and all checks are passing 🙌

File: python-package/lightgbm/compat.py
Comment (StrikerRUS) [inline python-package/lightgbm/compat.py:157]: Sorry, missed this moment in my previous review. Why dummy function and not simply `None` as for all other functions in this module?
https://github.com/microsoft/LightGBM/blob/e36cc9c1717665963a7fed09a5413a2bf5f0f50a/python-package/lightgbm/compat.py#L113-L122
https://github.com/microsoft/LightGBM/blob/e36cc9c1717665963a7fed09a5413a2bf5f0f50a/python-package/lightgbm/compat.py#L135-L137

Comment (jameslamb) [reply python-package/lightgbm/compat.py:157]: oh! No reason, I just didn't notice those. I'll update that here.

Comment (jameslamb) [reply python-package/lightgbm/compat.py]: updated in https://github.com/microsoft/LightGBM/pull/4378/commits/e4fed6f1300983993babe32111325ca98c448e5d

**Review by StrikerRUS — APPROVED** [review]
LGTM!

**Pull request merged**
